Accessibility (#54)
reactjs/fr.reactjs.org: content/docs/accessibility.md

- [Le WebAIM se penche sur la validation de formulaire](http://webaim.org/techniques/formvalidation/).

## Contrôle du focus {#focus-control}

Assurez-vous que votre application web peut être complètement utilisable avec le clavier seul :

- [Le WebAIM parle de l'accessibilité depuis le clavier](http://webaim.org/techniques/keyboard/).

### Focus par le clavier et contour de focus {#keyboard-focus-and-focus-outline}

Le focus par le clavier fait référence à l'élément courant dans le DOM qui est sélectionné pour accepter les saisies au clavier. Nous le voyons partout comme un contour tel que cela est montré sur l'image ci-dessous :

<img src="../images/docs/keyboard-focus.png" alt="Le contour bleu du focus clavier autour d'un lien sélectionné" />

N'utilisez que le CSS qui retire ce contour, par exemple en définissant `outline: 0`, si vous le remplacez par une autre implémentation de contour de focus.

### Mécanismes pour sauter au contenu désiré {#mechanisms-to-skip-to-desired-content}

Fournissez un mécanisme permettant aux utilisateurs de sauter les sections antérieures de navigation sur votre application, car cela facilite et accélère la navigation au clavier.

> Les liens de passage (*Skiplinks*) ou lien de navigation de passage sont des liens de navigation cachés qui ne sont visibles que lorsque l'utilisateur interagit au clavier avec la page. Ils sont très faciles à mettre en œuvre avec les ancres de pages internes et un peu de style :

Author: Merci pour la recommandation !

> - [WebAIM - Liens de navigation de passage](http://webaim.org/techniques/skipnav/).

Author: Ok, je l'ajouterai après intégration de tes remarques.

> Remarque :
>
> Bien qu'il s'agisse d'une caractéristique d'accessibilité très importante, c'est également une technique qui doit être utilisée à bon escient. Utilisez-la pour corriger le comportement du focus clavier lorsqu'il est perturbé, et non pour essayer d'anticiper la manière dont les utilisateurs souhaitent utiliser les applications.

## Événements de souris et de pointeur {#mouse-and-pointer-events}

Assurez-vous que toutes les fonctionnalités exposées via un événement de souris ou de pointeur sont également accessibles avec le clavier seul. Ne dépendre que du pointeur peut aboutir à de nombreuses situations où les utilisateurs de clavier ne pourront pas utiliser votre application.

> Pour illustrer cela, examinons un exemple prolifique où l'accessibilité est cassée par les événements de clics. Il s'agit du modèle de clic extérieur dans lequel un utilisateur peut désactiver un menu déroulant en cliquant à l'extérieur de l'élément.

Author: « Menu déroulant » n'est sans doute pas idéal, mais je trouve qu'« élément ouvert en surplomb » c'est assez lourd (comme le <strike>sur</strike>plomb) comme dénomination, non ?

> Pour illustrer cela, examinons un exemple prolifique où l'accessibilité est cassée par les événements de clics. Il s'agit du modèle de clic extérieur dans lequel un utilisateur peut désactiver un menu déroulant en cliquant à l'extérieur de l'élément.

Reviewer: En fait dans cet exemple précis, le *popover* simule bien une liste déroulante, donc on peut dire « liste déroulante »

> Pour illustrer cela, examinons un exemple prolifique où l'accessibilité est cassée par les événements de clics. Il s'agit du modèle de clic extérieur dans lequel un utilisateur peut désactiver un menu déroulant en cliquant à l'extérieur de l'élément.

Author: Changé en « liste déroulante »

}
```

Ce code expose la fonctionnalité aussi bien aux utilisateurs de dispositifs de pointage qu'aux utilisateurs de clavier. Remarquez également les propriétés ajoutées `aria-*` afin de supporter les lecteurs d'écran. Par souci de simplicité, les événements clavier permettant l'interaction avec les options du menu déroulant via les touches de direction n'ont pas été développés.

<img src="../images/docs/blur-popover-close.gif" alt="Une liste déroulante se fermant correctement pour les utilisateurs de souris et de clavier." />

> C'est un exemple des nombreux cas où le fait de ne dépendre que des événements de souris et de pointeur casse les fonctionnalités pour les utilisateurs de clavier. Toujours tester avec le clavier mettra immédiatement en évidence les problèmes qui peuvent ensuite être résolus à l'aide des gestionnaires d'événement clavier.

Reviewer 2: gestionnaires d'événement clavier. -> gestionnaires d'événements clavier.

## Widgets plus complexes {#more-complex-widgets}

> Une expérience utilisateur plus complexe ne doit pas signifier une expérience moins accessible. Alors que l'accessibilité est plus facile à réaliser en codant au plus près du HTML, même les widgets les plus complexes peuvent être codés de manière accessible.

Reviewer: ```suggestion
Une expérience utilisateur plus complexe ne doit pas signifier une expérience moins accessible. Alors que l'accessibilité est plus facile à réaliser en codant au plus près du HTML, même les éléments d'interface les plus complexes peuvent être codés de manière accessible.
```

Nous avons besoin ici de connaître les [rôles ARIA](https://www.w3.org/TR/wai-aria/#roles), ainsi que les [états et propriétés ARIA](https://www.w3.org/TR/wai-aria/#states_and_properties).

> Ce sont des boîtes à outils pleines d'attributs HTML pris en charge dans le JSX et qui nous permettent de construire des composants React entièrement accessibles et hautement fonctionnels.

Reviewer: ```suggestion
Ce sont des boîtes à outils pleines d'attributs HTML pris en charge par JSX et qui nous permettent de construire des composants React pleinement accessibles et hautement fonctionnels.
```

> Chaque type de widget a son modèle de design spécifique et devrait fonctionner de la même manière avec les utilisateurs et les agents utilisateurs :

Reviewer: ```suggestion
Chaque type d'élément d'interface a son modèle de conception spécifique et devrait fonctionner de la même manière avec les utilisateurs et les agents utilisateurs (notamment les navigateurs) :
```

> - [Pratique de création WAI-ARIA - Modèles de design et widgets](https://www.w3.org/TR/wai-aria-practices/#aria_ex) (*WIA* signifie *Web Accessibility Initiative* - NdT).

Reviewer: ```suggestion
- [Pratiques de création WAI-ARIA - Modèles de conception et éléments d’interface](https://www.w3.org/TR/wai-aria-practices/#aria_ex)
```

On a déjà pris soin d'élaborer WAI-ARIA plus haut, inutile de répéter.

- [Heydon Pickering - Exemples ARIA](http://heydonworks.com/practical_aria_examples/).
- [Composants inclusifs](https://inclusive-components.design/).

> ## Autres points de considération {#other-points-for-consideration}

Reviewer: ```suggestion
## Autres points à considérer {#other-points-for-consideration}
```

### Définir la langue {#setting-the-language}

Indiquez la langue des pages de texte de façon à ce que les logiciels de lecture d'écran puissent sélectionner les paramètres de voix appropriés :

> - [WebAIM - Document sur les langues](http://webaim.org/techniques/screenreader/#language).

Reviewer: ```suggestion
- [WebAIM - Langue du document](http://webaim.org/techniques/screenreader/#language).
```

Ahem

### Définir le titre du document {#setting-the-document-title}

Définissez la balise `<title>` du document pour décrire correctement le contenu de la page courante, afin de garantir que l'utilisateur soit au courant du contexte de la page en cours :

- [WCAG - Comprendre l'exigence du titre du document](https://www.w3.org/TR/UNDERSTANDING-WCAG20/navigation-mechanisms-title.html).

Nous pouvons le définir dans React en utilisant le [composant React Document Title](https://github.com/gaearon/react-document-title).

### Contraste des couleurs {#color-contrast}

Assurez-vous que tous les textes lisibles sur votre site web ont un contraste des couleurs suffisant pour rester lisibles de manière optimale par les utilisateurs malvoyants :

- [WCAG - Comprendre l'exigence du contraste des couleurs](https://www.w3.org/TR/UNDERSTANDING-WCAG20/visual-audio-contrast-contrast.html).

> - [Tout sur le contraste des couleurs et pourquoi vous devez le repenser](https://www.smashingmagazine.com/2014/10/color-contrast-tips-and-tools-for-accessibility/).

Reviewer: ```suggestion
- [Tout savoir sur le contraste des couleurs et pourquoi vous devriez le repenser](https://www.smashingmagazine.com/2014/10/color-contrast-tips-and-tools-for-accessibility/).
```

> - [A11yProject - Qu'est-ce qu'un contraste des couleurs ?](http://a11yproject.com/posts/what-is-color-contrast/)

Reviewer: ```suggestion
- [A11yProject - Qu'est-ce que le contraste des couleurs ?](http://a11yproject.com/posts/what-is-color-contrast/)
```

Il peut être fastidieux de calculer manuellement les combinaisons de couleurs appropriées pour toutes les situations sur votre site web, aussi vous est-il possible de [calculer une palette entière de couleurs accessible avec Colorable](http://jxnblk.com/colorable/).

Les outils aXe et WAVE mentionnés ci-dessous incluent également des tests de contraste de couleur et signalent les erreurs de contraste.

Si vous souhaitez étendre vos capacités de test de contraste, vous pouvez utiliser ces outils :

> - [WebAIM - Vérification de contraste de couleur](http://webaim.org/resources/contrastchecker/).

Reviewer 2: couleurs

Author: J'ai changé en "contraste de**s** couleur**s**", au moins le doute n'est plus permis 😅

> - [The Paciello Group - Analyseur de contraste de couleur](https://www.paciellogroup.com/resources/contrastanalyser/).

Reviewer 2: couleurs

> ## Développement et outils de test {#development-and-testing-tools}

Reviewer: ```suggestion
## Outils de développement et de test {#development-and-testing-tools}
```

Il existe de nombreux outils que nous pouvons utiliser pour nous assister durant la création d'applications web accessibles.

### Le clavier {#the-keyboard}

> De loin la vérification plus simple, mais aussi l'une des plus importantes, consiste à tester si l'ensemble de votre site web est accessible et utilisable avec le clavier seul. Procédez ainsi :

Reviewer 2: la vérification plus simple -> la vérification la plus simple

> De loin la vérification la plus simple, mais aussi l'une des plus importantes, consiste à tester si l'ensemble de votre site web est accessible et utilisable avec le clavier seul. Procédez ainsi :

Reviewer: ```suggestion
La vérification de loin la plus simple, mais aussi l'une des plus importantes, consiste à tester si l'ensemble de votre site web est accessible et utilisable avec le clavier seul. Procédez ainsi :
```

1. Débranchez votre souris.
1. Utilisez `Tab` et `Shift + Tab` pour naviguer.
1. Utilisez `Entrée` pour activer des éléments.

> 1. Le cas échéant, utilisez les touches de direction du clavier pour interagir avec certains éléments, tels que les menus et les menus déroulants.

Reviewer: ```suggestion
1. Le cas échéant, utilisez les touches de curseur du clavier pour interagir avec certains éléments, tels que les menus et les listes déroulantes.
```

### Assistance au développement {#development-assistance}

> Nous pouvons tester certaines fonctionnalités d'accessibilités directement dans notre code JSX. Souvent des contrôles intellisense sont déjà présents dans les IDE pour prendre en charge les rôles, états et propriétés ARIA. Nous avons également accès à l'outil suivant :

Reviewer: ```suggestion
Nous pouvons tester certaines fonctionnalités d'accessibilité directement dans notre code JSX. Souvent des contrôles automatiques sont déjà présents dans les EDI qui prennent en charge JSX pour vérifier les rôles, états et propriétés ARIA. Nous avons également accès à l'outil suivant :
```

#### eslint-plugin-jsx-a11y {#eslint-plugin-jsx-a11y}

> L'extension [eslint-plugin-jsx-a11y](https://github.com/evcohen/eslint-plugin-jsx-a11y) pour ESLint fournit des informations sur l'AST concernant les problèmes d'accessibilité dans votre JSX. De nombreux IDE vous permettent d'intégrer ces résultats directement dans des fenêtres d'analyse de code ou de code source.

Reviewer: ```suggestion
L'extension [eslint-plugin-jsx-a11y](https://github.com/evcohen/eslint-plugin-jsx-a11y) pour ESLint fournit des informations concernant les problèmes d'accessibilité dans votre JSX. De nombreux EDI vous permettent d'intégrer ces résultats directement dans leurs fenêtres d'analyse de code ou de code source.
```

On peut virer la mention de l'AST, là, ça n'a rien à faire ici, n'apporte rien, et la majorité des lecteurs ne sauront pas ce que c'est. Sans doute à proposer pour la VO aussi.

> [Create React App](https://github.com/facebookincubator/create-react-app) dispose de cette extension avec un sous-ensemble de règles activées. Si vous souhaitez activer encore plus de règles d'accessibilité, vous pouvez créer un fichier `.eslintrc` à la racine de votre projet avec ce contenu :

Reviewer: ```suggestion
[Create React App](https://facebook.github.io/create-react-app/) pré-configure cette extension avec un sous-ensemble de règles activées. Si vous souhaitez activer encore plus de règles d'accessibilité, vous pouvez créer un fichier `.eslintrc` à la racine de votre projet avec ce contenu :
```

Autant utiliser la nouvelle URL maintenant qu'ils ont un beau site 😉

  }
  ```

### Tester l'accessibilité dans le navigateur {#testing-accessibility-in-the-browser}

> Il existe un certain nombres d'outils pouvant exécuter des audits d'accessibilité sur des pages web dans votre navigateur. Veuillez les utiliser conjointement avec d'autres contrôles d'accessibilité mentionnés ici car ils ne peuvent que tester l'accessibilité technique de votre HTML.

Reviewer: ```suggestion
Il existe un certain nombre d'outils pour exécuter des audits d'accessibilité sur des pages web dans votre navigateur. Veuillez les utiliser conjointement avec d'autres contrôles d'accessibilité mentionnés ici car ils ne peuvent tester que l'accessibilité technique de votre HTML.
```

#### aXe, aXe-core et react-axe {#axe-axe-core-and-react-axe}

Deque Systems propose [aXe-core](https://github.com/dequelabs/axe-core) pour automatiser les tests d'accessibilité de bout en bout de vos applications. Ce module comprend des intégrations pour Selenium.

> [Le moteur d'accessibilité](https://www.deque.com/products/axe/) ou aXe, est une extension du navigateur d'inspection de l'accessibilité construite sur `aXe-core`.

Reviewer: ```suggestion
[*The Accessibility Engine*](https://www.deque.com/products/axe/) ou aXe, est une extension du navigateur qui fournit un inspecteur d'accessibilité en se basant sur `aXe-core`.
```

C'est un « nom propre », on ne traduit pas. Par ailleurs, c'était peu clair ta VF là 😁

> Vous pouvez aussi utiliser le module [react-axe](https://github.com/dylanb/react-axe) pour signaler directement ces résultats d'accessibilité dans la console durant le développement et le déboguage.

Reviewer: ```suggestion
Vous pouvez aussi utiliser le module [react-axe](https://github.com/dylanb/react-axe) pour signaler directement ces résultats d'accessibilité dans la console durant le développement et le débogage.
```

#### WebAIM WAVE {#webaim-wave}

L'[outil Web Accessibility Evaluation](http://wave.webaim.org/extension/) est une autre extension du navigateur pour l'accessibilité.

#### Inspecteurs d'accessibilité et arbre d'accessibilité {#accessibility-inspectors-and-the-accessibility-tree}

[L'arbre d'accessibilité](https://www.paciellogroup.com/blog/2015/01/the-browser-accessibility-tree/) est un sous-ensemble de l'arbre DOM qui contient des objets d'accessibilité pour chaque élément du DOM devant être exposé aux technologies d'assistance, telles que les lecteurs d'écrans.

Sur certains navigateurs, nous pouvons facilement consulter les informations d'accessibilité pour chaque élément de l'arbre d'accessibilité :

> - [En utilisant l'inspecteur d'accessibilité de Firefox](https://developer.mozilla.org/en-US/docs/Tools/Accessibility_inspector).

Reviewer: ```suggestion
- [Utiliser l'inspecteur d'accessibilité de Firefox](https://developer.mozilla.org/en-US/docs/Tools/Accessibility_inspector).
```

> - [En activant l'inspecteur d'accessibilité de Chrome](https://gist.github.com/marcysutton/0a42f815878c159517a55e6652e3b23a).

Reviewer: ```suggestion
- [Activer l'inspecteur d'accessibilité de Chrome](https://gist.github.com/marcysutton/0a42f815878c159517a55e6652e3b23a).
```

> - [En utilisant l'inspecteur d'accessibilité d'OS X Safari](https://developer.apple.com/library/content/documentation/Accessibility/Conceptual/AccessibilityMacOSX/OSXAXTestingApps.html).

Reviewer: ```suggestion
- [Utiliser l'inspecteur d'accessibilité d'OS X Safari](https://developer.apple.com/library/content/documentation/Accessibility/Conceptual/AccessibilityMacOSX/OSXAXTestingApps.html).
```

### Lecteurs d'écran {#screen-readers}

Tester avec un lecteur d'écran devrait faire partie de vos tests d'accessibilité.

Veuillez noter que les combinaisons navigateur / lecteur d'écran ont leur importance. Il est recommandé de tester votre application dans le navigateur le plus adapté au lecteur d'écran de votre choix.

> ### Lecteurs d'écran utilisés fréquemment {#commonly-used-screen-readers}

Reviewer: ```suggestion
### Lecteurs d'écran fréquemment utilisés {#commonly-used-screen-readers}
```

#### NVDA dans Firefox {#nvda-in-firefox}

> [*NonVisual Desktop Access*](https://www.nvaccess.org/) ou NVDA est un lecteur d'écran open source qui est largement utilisé.

Reviewer: ```suggestion
[*NonVisual Desktop Access*](https://www.nvaccess.org/) ou NVDA est un logiciel libre de lecteur d'écran qui est largement utilisé.
```

Reportez-vous aux guides suivants pour savoir comment utiliser au mieux NVDA :

- [WebAIM - Utiliser NVDA pour évaluer l'accessibilité web](http://webaim.org/articles/nvda/).
- [Deque - Raccourcis clavier NVDA](https://dequeuniversity.com/screenreaders/nvda-keyboard-shortcuts).

#### VoiceOver dans Safari {#voiceover-in-safari}

VoiceOver est un lecteur d'écran intégré dans les appareils d'Apple.

Reportez-vous aux guides suivants pour savoir comment activer et utiliser VoiceOver :

- [WebAIM - Utiliser VoiceOver pour évaluer l'accessibilité web](http://webaim.org/articles/voiceover/).
- [Deque - Raccourcis clavier pour VoiceOver sur OS X](https://dequeuniversity.com/screenreaders/voiceover-keyboard-shortcuts).
- [Deque - Raccourcis pour VoiceOver sur iOS](https://dequeuniversity.com/screenreaders/voiceover-ios-shortcuts).

#### JAWS dans Internet Explorer {#jaws-in-internet-explorer}

> [*Job Access With Speech*](http://www.freedomscientific.com/Products/Blindness/JAWS) ou JAWS, est un lecteur d'écran utilisé de manière prolifique sur Windows.

Reviewer: ```suggestion
[*Job Access With Speech*](http://www.freedomscientific.com/Products/Blindness/JAWS) ou JAWS, est un lecteur d'écran très largement utilisé sur Windows.
```

Reportez-vous aux guides suivants pour utiliser au mieux JAWS :

- [WebAIM - Utiliser JAWS pour évaluer l'accessibilité web](http://webaim.org/articles/jaws/).
- [Deque - Raccourcis clavier pour JAWS](https://dequeuniversity.com/screenreaders/jaws-keyboard-shortcuts).

### Autres lecteurs d'écran {#other-screen-readers}

#### ChromeVox dans Google Chrome {#chromevox-in-google-chrome}

[ChromeVox](http://www.chromevox.com/) est un lecteur d'écran intégré aux Chromebooks et est disponible [en tant qu'extension](https://chrome.google.com/webstore/detail/chromevox/kgejglhpjiefppelpmljglcjbhoiplfn?hl=fr) pour Google Chrome.

Reportez-vous aux guides suivants pour utiliser au mieux ChromeVox :

- [Aide de Google Chromebook - Utiliser le lecteur d'écran intégré](https://support.google.com/chromebook/answer/7031755?hl=fr).

> - [Références des raccourcis clavier pour ChromeVox Classic](http://www.chromevox.com/keyboard_shortcuts.html).

Reviewer: ```suggestion
- [Référence des raccourcis clavier pour ChromeVox Classic](http://www.chromevox.com/keyboard_shortcuts.html).
```
